Add CUTLASS backend support for torch.float8_e5m2 dtype

gaurav-redhat opened the pull request:
Add CUTLASS backend support for torch.float8_e5m2 dtype

Fixes assertion error in torch._scaled_mm with mixed FP8 dtypes during
FP8 training backward pass. When using torch.compile with CUTLASS backend,
torch._scaled_mm failed when called with mixed FP8 dtypes (float8_e4m3fn
and float8_e5m2).

The CUTLASS backend was missing support for torch.float8_e5m2. This adds:

CUTLASS backend (cutlass_utils.py):
- Added torch.float8_e5m2 to XW_DTYPES (supported input types)
- Added CUDA type mapping (__nv_fp8_e5m2)
- Added CUTLASS library DataType mapping (DataType.e5m2)
- Added dtype matching logic for e5m2
- Added accumulator dtype inference (uses float32 like e4m3fn)
- Added alignment support [16, 8, 4, 2] elements
- Fixed get_accumulator_dtype to handle mixed FP8 types

TMA compatibility (utils.py):
- Added torch.float8_e5m2 to TMA-compatible dtypes check
- Added e5m2 to FP8 special case requiring inner dimension >= 32

This enables mixed FP8 precision operations (e4m3fn x e5m2) in
CUTLASS-accelerated kernels, particularly for FP8 training backward passes.

Fixes #171165

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

gaurav-redhat committed:
Add CUTLASS backend support for torch.float8_e5m2 dtype

Fixes assertion error in torch._scaled_mm with mixed FP8 dtypes during
FP8 training backward pass. When using torch.compile with CUTLASS backend,
torch._scaled_mm failed when called with mixed FP8 dtypes (float8_e4m3fn
and float8_e5m2).

The CUTLASS backend was missing support for torch.float8_e5m2. This adds:

CUTLASS backend (cutlass_utils.py):
- Added torch.float8_e5m2 to XW_DTYPES (supported input types)
- Added CUDA type mapping (__nv_fp8_e5m2)
- Added CUTLASS library DataType mapping (DataType.e5m2)
- Added dtype matching logic for e5m2
- Added accumulator dtype inference (uses float32 like e4m3fn)
- Added alignment support [16, 8, 4, 2] elements
- Fixed get_accumulator_dtype to handle mixed FP8 types

TMA compatibility (utils.py):
- Added torch.float8_e5m2 to TMA-compatible dtypes check
- Added e5m2 to FP8 special case requiring inner dimension >= 32

This enables mixed FP8 precision operations (e4m3fn x e5m2) in
CUTLASS-accelerated kernels, particularly for FP8 training backward passes.

Fixes #171165

drisspg reviewed (commented):
can you add a test please :)

gaurav-redhat committed:
Add test for mixed FP8 scaled_mm with CUTLASS

gaurav-redhat committed:
fix: remove dead code in get_accumulator_dtype

The mixed FP8 case is already handled by early return at lines 373-376,
so the fp8_types check at lines 382-384 was unreachable dead code.

eellison reviewed (commented):
looks good but deferring to @mlazos and @drisspg  - can you add a new test?

pytorchmergebot committed:
Apply lint patches

Generated by apply-lint bot, see https://github.com/pytorch/pytorch/actions/runs/23501409730
